### DISCUSSION TITLE
[Feature] Default to x64-windows triplet on x64 Windows

### DISCUSSION THREAD

**Discussion by VelocityRa** [microsoft/vcpkg#17254] [Ideas]
Most people don't use 32-bit builds these days, so we have to always specify the triplet which is cumbersome, or set the env-var which is not elegant.

### Implementation
Instead of this:
https://github.com/microsoft/vcpkg/blob/22c8e3a23afc6be45020cd3f8c92bcea783ce8fa/toolsrc/src/vcpkg/triplet.cpp#L100
there should be a check to see if the host is 64-bit and use the `x64-windows` triplet if so.

`IsWow64Process` or `IsWow64Process2` can be used for this (will work even if vcpkg is compiled as 32-bit).

**Comment by PhoebeHui** [comment]
Duplicated to issue https://github.com/microsoft/vcpkg/issues/1254 

Related PR https://github.com/microsoft/vcpkg/pull/11610

**Comment by cloudhan** [comment]
if I have commit permission, I will change
https://github.com/microsoft/vcpkg/blob/22c8e3a23afc6be45020cd3f8c92bcea783ce8fa/toolsrc/src/vcpkg/triplet.cpp#L99-L100
to
```cpp
#if defined(_WIN64)
                return Triplet::X64_WINDOWS;
#elif defined(_WIN32)
                return Triplet::X86_WINDOWS;
```
right now and force push it...

See how many thumb ups in #1254

**Comment by VelocityRa** [comment]
@cloudhan This is wrong, `_WIN64` will only check if vcpkg is compiled in a 64-bit environment, not if the host is a 64-bit operating system. If vcpkg itself was compiled as 32-bit on a 64-bit system it wouldn't work.

See my implementation recommendation above instead.

**Comment by cloudhan** [comment]
@VelocityRa Good point, and seems the bootstrap script indeed build 32bit version of vcpkg.exe on a 64bit platform. Another feature request, here you go.

**Comment by MarkIngramUK** [comment]
This has just caught me out on a fresh install (again). 64-bit by default on 64-bit machines seems like a sensible choice.

**Comment by PhoebeHui** [comment]
Same issue https://github.com/microsoft/vcpkg/issues/3882 and https://github.com/microsoft/vcpkg/issues/1254

**Comment by VelocityRa** [comment]
Why was the issue closed? A note in the README it is not a fix to the actual problem.

**Reply by PhoebeHui** [reply]
@VelocityRa, we doesn't close the thread. if you take a look at the adderss, you will noted that we just moved it to discussions. And in future, we would move other common topics to discussions.

**Reply by VelocityRa** [reply]
Yes, but why?
There's 1.3k open Issues, what made this one particularly fit for being a "Discussion" instead of an Issue.
There is not much discussion to be done anyway... the fix seems pretty straightforward.

**Reply by PhoebeHui** [reply]
@VelocityRa, sorry for late response. This is not about how to fix it. It’s a history topic, and we have serveral issues (eg: #3882 #1254) about it, we haven’t accept it before by some reasons, however, this may be changed in future, so I moved it to discussion.

**Comment by dg0yt** [comment]
The strangest thing in my experience:
The target defaults to `x86-windows` while the host defaults to `x64-windows`.  E.g. for `vcpkg install fmt`:
~~~
    fmt[core]:x86-windows -> 7.1.3#3
  * vcpkg-cmake[core]:x64-windows -> 2021-02-28#1
  * vcpkg-cmake-config[core]:x64-windows -> 2021-02-26#1
~~~

**Reply by autoantwort** [reply]
The default host triplets is determined by looking at the host system ([here](https://github.com/microsoft/vcpkg-tool/blob/af0237831437d5a6ed4d17fc4f780e27a84e3c8e/src/vcpkg/triplet.cpp#L76)), but the default triplet is hardcoded to `x86-windows` on WIN32 ([here](https://github.com/microsoft/vcpkg-tool/blob/af0237831437d5a6ed4d17fc4f780e27a84e3c8e/src/vcpkg/triplet.cpp#L102)). I don't know why they are reluctant to change this.

**Comment by SlySven** [comment]
> there should be a check to see if the host is 64-bit and use the x64-windows triplet if so.

FTR From a MSYS2/Ming-w64 point of view they dropped supporting 32-bit *hosts* over a year ago - however their toolchains are available for both 32-bit and 64-bit *targets*. Obviously, due to [Windows-On-Windows64](https://docs.microsoft.com/en-us/windows/win32/winprog64/wow64-implementation-details) a 32-bit product compiled by that toolchain should run (with the inherit limitations that that number of bits imposes) on both architectures whereas a 64-bit one won't...

**Comment by Whotakesmyname** [comment]
This is the second most impressive x86 default configuration I have ever come across. And the 1st place belongs to Microsoft Office.

Compatibility hurts.

**Comment by AugP** [comment]
Bumping this to state that in September we will be changing the default triplet to a detected host system triplet (which on x64 devices will be x64). For now, there is a warning message so users can prepare in case their scripts are calling vcpkg and are assuming the current default. More details: https://devblogs.microsoft.com/cppblog/vcpkg-2023-02-24-release-default-triplet-changes-risc-v-support-bug-fixes-and-more/

**Reply by AugP** [reply]
Also updating that we did indeed ship this update :)

**Comment by vicroms** [comment]
The default has been changed since [version 2023-10-19](https://devblogs.microsoft.com/cppblog/vcpkg-2023-10-19-release-export-for-manifests-documentation-improvements-and-more/).
